plugin/pulpcore/plugin/stages/declarative_version.py

```diff
@@ -0,0 +1,146 @@
+import asyncio
+from gettext import gettext as _
+
+from pulpcore.plugin.models import RepositoryVersion
+from pulpcore.plugin.tasking import WorkingDirectory
+
+from .api import create_pipeline, end_stage
+from .artifact_stages import artifact_downloader, artifact_saver, query_existing_artifacts
+from .association_stages import content_unit_association, content_unit_unassociation
+from .content_unit_stages import content_unit_saver, query_existing_content_units
+
+
+class FirstStage:
```

Comment:
Can you help me understand the value of this interface?

By adopting the *Command* design pattern recommendation and requiring each stage to simply be a *callable* by implementing `__call__()` this interface could no longer be necessary.  Stages being an object is useful in that it supports separate stage construction/initialization and usage.

Reply:
This interface came out of [Simon's feedback](https://github.com/pulp/pulp_file/commit/c23bec069e12a485000f048e486e1d89cce65beb#r29701942). For all the reasons he says, the plugin writer will need to make an object like this one. I think the value of providing it here is that we can clearly prompt the plugin writer to build the right thing because we can say they need to subclass this.

The only other option would be for them to make the object just like this on its own, and pass the function that returns the co-routine. This is tricky because if you call `MyPluginFirstStage(myarg=True).my_coroutine()` it won't work with `create_pipeline()` but if you call `MyPluginFirstStage(myarg=True).my_coroutine` it will. I don't think this is very intuative to people.

Reply:
Actually, I think it makes it more consistent and intuitive for plugin writers.  They would pass a list any *callable* as a stage to create a pipeline and in some cases it's a *function* and others it's a *callable* object.  Looking at `create_pipeline()` it calls each stage passing the queues.

Reply:
After thinking about this more I agree w/ you. I'm going to switch the `stage` to be `__call__`. I'll do the same w/ `FirstStage` as well.